### ISSUE TITLE
Allow promotion using the raw type

### ISSUE THREAD

**Issue by karlklose** [dart-lang/sdk#17281]
Consider this example:

```dart
class B<T> {}

class C<T> extends B<T> {
  var foo;
  bar(B<T> b) {
    if (b is C &amp;&amp; b.foo) {}
  }
}
```

It would be really useful if `b` could be promoted to `C<T>` (even `C<dynamic>` would work for this example, but not in general).

In general, this kind of promotion is not possible, but we should support it for the simple, useful cases like this.

There are three cases here, depending on the supertype relation:

1. The type arguments to the subtype can be inferred completely from the supertype. This is the case we should support.
2. We would have to guess some type arguments, like in `C<S, T> extends B<T>`.
3. The promotion would only be true for certain instantiations, like in `C<T> extends B<List<T>>`.

In cases 2 and 3 we should not promote, but should hint that promotion would work if the type arguments were provided.

**Comment by eernstg** [comment]
**tl;dr**  We might do this, but there is no decision yet.  **rd;lt**

The language team has discussed adding support for type inference in situations where a raw type is given and the context provides some kind of constraints on the "reasonable" type arguments. In particular, we have a proposed [feature spec](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/mixin-inference.md) for doing this with raw classes used as mixins.

A similar approach could be used in other situations, e.g., a raw `C` could have type arguments inferred (in some situations) for an expression like `x is C`. That is, if possible, inference would be used to select type arguments `T1..Tk` such that `C<T1..Tk>` is a subtype of the statically known type of `x`, and `x` is then known to have type `C<T1..Tk>` in code where that test is known to have yielded `true`. It may or may not be useful to treat `x is! C` similarly, and it may or may not be a good choice to use [instantiate-to-bound](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/instantiate-to-bound.md) when inference fails (we might also make it an error). So we surely still need to discuss a number of things, and we may or may not end up doing this&mdash;maybe it's "smart" and "convenient" at first, but if it causes code to be so much less readable that the overall effect is expected to be negative then we might not do it at all.

Moreover, we certainly do not have this feature implemented at this point:

```dart
import 'package:expect/expect.dart';

class B<X> {}

class C<X> extends B<X> {
  X foo;
  C(this.foo);
}

bool bar (B<int> b) => b is C && b.foo.isEven;

main() {
  Expect.isTrue(bar(new C<int>(42)));
}
```

Running `tools/test.py` in various configurations on the above program and grepping the output for `FAILED` yields this:
```
FAILED: dart2analyzer-none release_x64 language_2/scratch_test
FAILED: dart2analyzer-none release_x64 language_2/scratch_test
FAILED: dart2js-d8 release_x64 language_2/scratch_test
FAILED: dart2js-d8-checked release_x64 language_2/scratch_test
FAILED: dart2js-d8 release_x64 language_2/scratch_test
FAILED: dartk-vm release_x64 language_2/scratch_test
FAILED: dartk-vm-checked release_x64 language_2/scratch_test
FAILED: dartkp-dart_precompiled release_x64 language_2/scratch_test
FAILED: dartkp-dart_precompiled-checked release_x64 language_2/scratch_test
FAILED: dartdevc-chrome release_x64 language_2/scratch_test
FAILED: dartdevc-chrome-checked release_x64 language_2/scratch_test
FAILED: dartdevk-chrome release_x64 language_2/scratch_test
FAILED: dartdevk-chrome-checked release_x64 language_2/scratch_test
```
A closer inspection shows that we get "foo isn't defined" for all failures, i.e., `b` was not promoted.

So I believe the right thing to do is to reopen this issue.
